ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/BlockPropagationManager.java

```diff
@@ -397,34 +403,110 @@ private CompletableFuture<Block> retrieveParentBlock(final BlockHeader blockHead
+              if (throwable != null) {
+                LOG.warn(
+                    "Failed to retrieve block "
+                        + logBlockNumberMaybeHash(blockNumber, maybeBlockHash),
+                    throwable);
+              }
+              requestedNonAnnouncedBlocks.remove(blockNumber);
+              if (block != null) {
+                debugLambda(LOG, "Successfully retrieved block {}", block::toLogString);
+                maybeBlockHash.ifPresentOrElse(
+                    requestedBlocks::remove,
+                    () ->
+                        // in case we successfully retrieved only by block number, when can remove
+                        // the request by hash too
+                        requestedBlocks.remove(block.getHash()));
+              } else {
+                // this could happen if we give up at some point since we find that it make no sense
+                // to retry
+                debugLambda(LOG, "Block {} not retrieved", block::toLogString);
+              }
+            });
+  }
+
+  private CompletableFuture<Block> repeatableGetBlockFromPeer(
+      final Optional<EthPeer> preferredPeer,
+      final long blockNumber,
+      final Optional<Hash> maybeBlockHash) {
+    return exceptionallyCompose(
+            scheduleGetBlockFromPeers(preferredPeer, blockNumber, maybeBlockHash),
+            handleGetBlockErrors(blockNumber, maybeBlockHash))
+        .thenCompose(r -> maybeRepeatGetBlock(blockNumber, maybeBlockHash));
+  }
+
+  private Function<Throwable, CompletionStage<Block>> handleGetBlockErrors(
+      final long blockNumber, final Optional<Hash> maybeBlockHash) {
+    return throwable -> {
+      debugLambda(
+          LOG,
+          "Temporary failure retrieving block {} from peers with error {}",
+          () -> logBlockNumberMaybeHash(blockNumber, maybeBlockHash),
+          throwable::toString);
+      return CompletableFuture.completedFuture(null);
+    };
+  }
+
+  private CompletableFuture<Block> maybeRepeatGetBlock(
+      final long blockNumber, final Optional<Hash> maybeBlockHash) {
+    final MutableBlockchain blockchain = protocolContext.getBlockchain();
+    final Optional<Block> maybeBlock =
+        maybeBlockHash
+            .map(hash -> blockchain.getBlockByHash(hash))
+            .orElseGet(() -> blockchain.getBlockByNumber(blockNumber));
+
+    // check if we got this block by other means
+    if (maybeBlock.isPresent()) {
+      final Block block = maybeBlock.get();
+      debugLambda(
+          LOG, "No need to retry to get block {} since it is already present", block::toLogString);
+      return CompletableFuture.completedFuture(block);
+    }
+
+    long distanceFromChainHead = blockNumber - blockchain.getChainHeadBlockNumber();
+    if (!config.getBlockPropagationRange().contains(distanceFromChainHead)) {
+      debugLambda(
+          LOG,
+          "Not retrying to get block {} since we are too far from local chain head {}",
```

Comment:
Is this a good place to suggest the user to start over with a checkpoint sync? If we got here during a backwards sync, that might be the only way to recover.

Reply:
Not sure I got this, backward sync does not run along with BlockPropagationManager